Fixes #35642 - updated the message in the new UI to show the host's creator

nofaralfasi opened the pull request:
Updated the message in the new UI to show the owner of the host. 
The new message will show (for example): "Created 3 minutes ago owned by Nofar Alfassi (updated 3 minutes ago)".
Any suggestions regarding this change / the new syntax are welcome.

Before screenshot:
![image](https://user-images.githubusercontent.com/38184193/198251732-cba27897-ea21-41bf-96c3-786c61bba754.png)

After screenshot:
![image](https://user-images.githubusercontent.com/38184193/198253274-c2b2e549-a7e5-4f76-bf16-e236a5c3846f.png)

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

ekohl reviewed (commented):
Owner is a separate field that can be edited. Should it really be shown as part of this or a separate entry?

On the old host page we also used to have an icon that indicated if the owner was a group or a user. That's now not visible, which could also be addressed better if it was its own entry.

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/HostDetails/index.js:
```diff
@@ -216,7 +216,11 @@ const HostDetails = ({
+                    sprintf(
+                      __('Created %s owned by %s'),
+                      date,
+                      response.owner_name
```
I believe that the original issue was that we don't display the information about the user who created the host.
The owner of the host and the user who created the host are not the same users sometimes.
One option to retrieve that data is possible through Audits, e.g:
```rb
Audit.where(id: host.audit_ids, action: 'create').first.username
```

another option could be to save that data as `user_id` or `created_by_user` on the host model on creation, so it will be easier to retrieve that data.

so later in the UI I would expect something like:
```js
sprintf(__('Created %s by %s'), date, response.created_by_user)
```

ares replied:
:dart:

ekohl replied:
That sounds correct, though we should also consider cases where the information is unavailable. For example, what if the user is removed? Perhaps audit logs have been pruned. The host may also have been created via automation (sending in facts for example) in which case I think the active user is the admin user. I'm not sure we can properly distinguish that.

That's why I think the UI should somehow deal with no (known) creator. For example, omitting the field, showing "unavailable" or something else.

nofaralfasi replied:
Thanks for the response @ekohl!

nofaralfasi committed:
Fixes #35642 - update the message in the new UI to display the host's creator

Add `creator_id` property to the Host model to save the user who created each host.
This property is used in the new host details UI to display the user who created the host.

nofaralfasi commented:
**Some notes regarding this PR:** 
My main goal was to update the message in the new UI, to show the host's creator instead of the host's owner.
The UI will look the same, but now the user displayed in the message is the creator instead of the owner.

I added a new column named `created_by_user_id` to the `hosts` table.
I didn't add a reference to it (foreign_key or index), because it didn't seem necessary. 

Also, I added the new property (`created_by_user_id`) to the hosts API, but I wasn't sure if there is any documentation that I need to update. 

If anyone has other thoughts or suggestions, I would be happy to hear them :)

nofaralfasi commented:
[test unit]

Ron-Lavi commented on app/views/api/v2/hosts/main.json.rabl:
```diff
@@ -37,6 +37,13 @@ node :hostgroup_name do |host|
+node :created_by_user do |host|
+  if host.created_by_user_id.present?
+    user = User.find(host.created_by_user_id)
+    "by #{user.firstname} #{user.lastname}"
```
Works for me, any chance there's also `User.username`?

Ron-Lavi replied:
except for the "by" string that should be part of the frontend imho

nofaralfasi replied:
> Works for me, any chance there's also `User.username`?

I didn't find any shortcut for that (username didn't work).

ekohl reviewed (commented):
Should it be `creator` instead of `created_by_user`? I think we generally prefer a noun for these properties.

nofaralfasi commented:
> Should it be `creator` instead of `created_by_user`? I think we generally prefer a noun for these properties.

Sure, if it is clear that the creator is the user, then no problem.

ekohl commented on db/migrate/20221025151600_add_created_by_user_id_to_hosts.rb:
```diff
@@ -0,0 +1,5 @@
+class AddCreatedByUserIdToHosts < ActiveRecord::Migration[6.1]
+  def change
+    add_column :hosts, :created_by_user_id, :integer
```
I'd prefer if this was a proper foreign key to a user. If the user is deleted, it should be set to NULL because an ID doesn't really say anything. We have the audits table if people really want to know what happened anyway.

nofaralfasi replied:
Yeah, I wasn't sure what to decide regarding this.
Thanks for the comment, I'll add it.

nofaralfasi commented on webpack/assets/javascripts/react_app/components/HostDetails/index.js:
```diff
@@ -216,13 +216,11 @@ const HostDetails = ({
+                  {date => sprintf(__('Created %s'), date)}
+                </RelativeDateTime>
+                {response.creator
+                  ? sprintf(__(' by %s '), response.creator)
```
@Ron-Lavi, is my syntax here is ok? (with the spaces at the begging and at the end of the text)

MariaAga replied:
I would move both the spaces to the html `{' '}`

ekohl commented on app/views/api/v2/hosts/main.json.rabl:
```diff
@@ -37,6 +37,13 @@ node :hostgroup_name do |host|
 end
 
+node :creator do |host|
+  if host.creator_id.present?
+    user = User.find(host.creator_id)
+    "#{user.firstname} #{user.lastname}"
+  end
```
I think this can be simplified to:
```suggestion
  host.creator&.fullname
```

However, It may be valuable to make it a hash where you can also provide the ID, rather than just the name.

nofaralfasi replied:
Hi @ekohl, I wasn't sure what you mean by making it a hash (to display in the UI?), but I added the `creator_id` param to the params we send in the response.
Please let me know if you meant something else.

ekohl replied:
I meant that now you present the creator as a string, but the API is supposed to be generic. It's not just for the UI. So right now you return
```json
{
  "creator": "Some Name"
}
```

But what I was suggesting is:
```json
{
  "creator": {
    "id": 1234,
    "fullname": "Some Name"
  }
}
```

That way you can expand it in the future. You can also create a link to the user or search for all hosts created by the user (searching using names is a bad idea here, ID is much more reliable).

And if there is no creator, it should be:
```json
{
  "creator": null
}
```

nofaralfasi replied:
I got what you mean. 
But for example for the host's owner, we use `owner_id` and `owner_name` for the API and only display the `owner_name` in the UI. Isn't that enough? Or is it better to present it as a hash for future uses (like expanding it)?

ekohl replied:
I think it's a good question. In retrospect (and if you're design an API v3) I'd change owner to a hash, just like I suggested for creator. In that case I'd even add a field `href` with `/api/v3/users/:id` as the value so clients don't need to guess the URLs.

However, an API v3 is without doubt out of scope here. The question is then what we should do now. A few options:

* remain consistent and add `creator_name` (or `creator_fullname`) and `creator_id`.
* Implement what I suggested above
* Do both: add `creator`, `creator_name` and `creator_id`

This is something where I'd get more feedback from others. You could open a post with a poll on https://community.theforeman.org/c/development/9 or even as an RFC. Others, such as the authors of https://github.com/theforeman/foreman-ansible-modules (large user of the API), can weigh in.

nofaralfasi replied:
Thank you Ewoud for all the help and for that clear and detailed answer!

nofaralfasi replied:
I opened this RFC [post](https://community.theforeman.org/t/rfc-adding-hosts-creator-field/31231) to see what others think about it.

nofaralfasi replied:
From the responses on the post, it seems that the first option (remain consistent and send two separate fields) is acceptable.
Please let me know if anything else is needed!

ekohl reviewed (commented):
Do be careful that this could create a n+1 query pattern where on the host list it may perform an additional DB query for every host (to get the creator). That's often killing for performance.

Also be sure that the user who performs the request actually has access to see the user. I'm not sure how we should deal with that but it's certainly something we always try to think about.

ekohl commented on app/models/host/base.rb:
```diff
@@ -348,6 +351,10 @@ def render_template(template:, **params)
       template.render(host: self, **params)
     end
 
+    def creator
```
I don't think this is needed nor desired. I'd expect that the `belongs_to` already creates it under the name creator. Or otherwise, you should pass the option (though I'd have to look it up myself how that's called).

nofaralfasi replied:
I know what you are talking about. I'll look into that. Thanks a lot!

nofaralfasi commented:
@ekohl do I need to add the `creator_id` property to the list here https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L71?
Or maybe it is unnecessary since it is generated according to the user?

ekohl commented:
It's a good question. I haven't really done that much with the API implementation so I'd try to ask someone who has. Though I was looking through the list of authors and it appears few people have. @adamruzicka perhaps you know?

nofaralfasi commented:
@ekohl sorry Ewoud, the review request was intended for Adam.

adamruzicka commented:
> do I need to add the creator_id property to the list here

I'd say no. The list describes which parameters the controller is able to accept. Because `creator_id` always gets overwritten, it would be probably confusing to market it as something the users can set explicitly.

nofaralfasi commented:
> I'd say no. The list describes which parameters the controller is able to accept. Because `creator_id` always gets overwritten, it would be probably confusing to market it as something the users can set explicitly.

And we don't want to allow to overwrite it?

adamruzicka commented:
Yeah, we don't want to allow it. Whoever creates the host is the creator and that's it.

nofaralfasi commented:
This PR is ready for review, there are no more changes on my side.

adamruzicka reviewed (approved):
This works well.

I was about to ask whether you considered filling in the association for already existing hosts from audits, but you already answered that in the [discourse thread](https://community.theforeman.org/t/rfc-adding-hosts-creator-field/31231/7?u=aruzicka).

Well done

adamruzicka commented:
Thank you @nofaralfasi !
